File: packages/core/src/identity.ts
// Identity = a self-owned Ed25519 keypair rendered as a did:key.
// did:key needs no authority and no blockchain: the DID *is* the public key,
// so anyone can verify a signature by parsing the DID. This is the whole
// identity layer for v0 (the "wallet custodies, the DID identifies" split).

import { generateKeyPairSync, createPublicKey, type KeyObject } from "node:crypto";
import { base58encode, base58decode } from "./base58";

// Multicodec prefix for an Ed25519 public key (0xed 0x01).
const ED25519_PREFIX = new Uint8Array([0xed, 0x01]);

export interface Identity {
  did: string;
  publicKey: KeyObject;
  privateKey: KeyObject;
}

function rawPublicKey(pub: KeyObject): Uint8Array {
  const jwk = pub.export({ format: "jwk" }) as { x: string };
  return new Uint8Array(Buffer.from(jwk.x, "base64url"));
}

export function didFromRawPublicKey(raw: Uint8Array): string {
  const prefixed = new Uint8Array(ED25519_PREFIX.length + raw.length);
  prefixed.set(ED25519_PREFIX, 0);
  prefixed.set(raw, ED25519_PREFIX.length);
  return "did:key:z" + base58encode(prefixed);
}

export function generateIdentity(): Identity {
  const { publicKey, privateKey } = generateKeyPairSync("ed25519");
  return { did: didFromRawPublicKey(rawPublicKey(publicKey)), publicKey, privateKey };
}

// Resolve a did:key back into a usable public key for signature verification.
export function publicKeyFromDid(did: string): KeyObject {
  if (!did.startsWith("did:key:z")) throw new Error("unsupported DID method: " + did);
  const decoded = base58decode(did.slice("did:key:z".length));
  if (decoded[0] !== 0xed || decoded[1] !== 0x01) {
    throw new Error("not an Ed25519 did:key");
  }
  const raw = decoded.slice(2);
  const x = Buffer.from(raw).toString("base64url");
  return createPublicKey({ key: { kty: "OKP", crv: "Ed25519", x }, format: "jwk" });
}

File: sdk/python/aleph_protocol/identity.py
"""did:key identity (Ed25519) — the DID *is* the public key. Matches
@aleph/core's identity.ts: multicodec prefix 0xed 0x01 + raw key, base58btc,
prefixed with 'z'."""

from dataclasses import dataclass

from cryptography.hazmat.primitives.asymmetric.ed25519 import (
    Ed25519PrivateKey,
    Ed25519PublicKey,
)
from cryptography.hazmat.primitives.serialization import Encoding, PublicFormat

from .base58 import b58decode, b58encode

_ED25519_PREFIX = bytes([0xED, 0x01])
_DIDKEY = "did:key:z"


@dataclass
class Identity:
    did: str
    private_key: Ed25519PrivateKey


def did_from_public_key(pub: Ed25519PublicKey) -> str:
    raw = pub.public_bytes(Encoding.Raw, PublicFormat.Raw)
    return _DIDKEY + b58encode(_ED25519_PREFIX + raw)


def generate_identity() -> Identity:
    priv = Ed25519PrivateKey.generate()
    return Identity(did=did_from_public_key(priv.public_key()), private_key=priv)


def public_key_from_did(did: str) -> Ed25519PublicKey:
    if not did.startswith(_DIDKEY):
        raise ValueError("unsupported DID method: " + did)
    decoded = b58decode(did[len(_DIDKEY) :])
    if len(decoded) < 2 or decoded[0] != 0xED or decoded[1] != 0x01:
        raise ValueError("not an Ed25519 did:key")
    return Ed25519PublicKey.from_public_bytes(bytes(decoded[2:]))
